src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ValidationHelpers.cs

```diff
@@ -57,49 +57,66 @@ public static string GetModelErrorMessageOrDefault(ModelError modelError)
                 {
                     return new[] { ms };
                 }
-
-                return Enumerable.Empty<ModelStateEntry>();
             }
-            else
+            else if (viewData.ModelState.Count > 0)
             {
                 var metadata = viewData.ModelMetadata;
-                var orderer = new ErrorsOrderer(metadata);
+                var modelStateDictionary = viewData.ModelState;
+                var entries = new List<ModelStateEntry>();
+                Visit(modelStateDictionary, modelStateDictionary.Root, metadata, entries);
+
+                Debug.Assert(entries[entries.Count - 1] == modelStateDictionary.Root);
+                // Special case the root node. It is only 'Visible` if it has any errors associated with it.
+                if (modelStateDictionary.Root.Errors.Count == 0)
+                {
+                    entries.RemoveAt(entries.Count - 1);
+                }
+
+                if (entries.Count != modelStateDictionary.Count)
+                {
+                    // Account for entries in the ModelStateDictionary that do not have corresponding ModelMetadata values.
+                    foreach (var entry in modelStateDictionary)
+                    {
+                        if (!entries.Contains(entry.Value))
+                        {
+                            entries.Add(entry.Value);
+                        }
+                    }
+                }
 
-                return viewData.ModelState
-                    .OrderBy(data => orderer.GetOrder(data.Key))
-                    .Select(ms => ms.Value);
+                return entries;
             }
+
+            return EmptyModelStateEntries;
         }
 
-        // Helper for sorting modelStates to respect the ordering in the metadata.
-        // ModelState doesn't refer to ModelMetadata, but we can correlate via the property name.
-        private class ErrorsOrderer
+        private static void Visit(
+            ModelStateDictionary dictionary,
+            ModelStateEntry modelStateEntry,
+            ModelMetadata metadata,
+            List<ModelStateEntry> orderedModelStateEntries)
         {
-            private readonly Dictionary<string, int> _ordering =
-                new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
-
-            public ErrorsOrderer(ModelMetadata metadata)
+            if (metadata.Properties.Count == 0)
             {
-                if (metadata == null)
-                {
-                    throw new ArgumentNullException(nameof(metadata));
-                }
-
-                foreach (var data in metadata.Properties)
+                var descendantEntries = dictionary.FindKeysWithPrefix(metadata.PropertyName);
```

Comment:
We had previously stated that we don't care about ordering within collections, let alone the collections themselves. It would be nice if it was correct, but not at the cost of overly complicated or slow code. Or taking too long to implement.